packages/ui-components/src/components/Map/LeafletMap.js

```diff
@@ -0,0 +1,281 @@
+/*
+ * Tupaia
+ *  Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+/**
+ * LeafletMap
+ *
+ * Wrapper around Leaflet, to control position and zoom via props
+ * while still allowing animation.
+ *
+ * Just using a bare react-leaflet map means that updating position/zoom/bounds
+ * props snaps to that new coordinate, but we want to animate! The way to do that
+ * is to get a ref to the leaflet map element and call `flyTo` and `flyToBounds`
+ * but that gets complex quickly. So, this component is here to encapsulate that
+ * complexity.
+ */
+
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+import { MapContainer as LeafletMapContainer } from 'react-leaflet';
+import { LeafletStyles } from './LeafletStyles';
+import { DEFAULT_BOUNDS } from './constants';
+
+const Map = styled(LeafletMapContainer)`
+  ${LeafletStyles};
+`;
+
+function arePositionsEqual(a, b) {
+  if (a && b) {
+    // if both centers are defined, equal if the position matches
+    return a[0] === b[0] && a[1] === b[1];
+  }
+  // if either are not defined, equal only if both are not defined
+  return a === b;
+}
+
+function areBoundsEqual(a, b) {
+  if (a && b) {
+    return arePositionsEqual(a[0], b[0]) && arePositionsEqual(a[1], b[1]);
+  }
+  // if either are not defined, equal only if both are not defined
+  return a === b;
+}
+
+function areBoundsValid(b) {
+  return Array.isArray(b) && b.length === 2;
+}
+export class LeafletMap extends Component {
```

Comment:
Would have thought there'd be a prettier rule about this, but should there be a space above the class definition?

Reply:
Yes agreed. No there's no rule around it though.